nanoDFT/sparse_symmetric_ERI.py

```diff
@@ -62,35 +80,153 @@ def num_repetitions_fast(value):
+            diff_JK    = diff_JK + jax.ops.segment_sum(dm_values, ss_indices, N**2) * (-HYB_B3LYP/2)**is_K_matrix # causes peak 2 
+            mask       = (mask-1)*jnp.sum(diff_JK) + mask 
+            return [diff_JK, mask]
+
+        batches = nonzero_indices.shape[0] # before pmap, tensor had shape (nipus, batches, -1) so [0]=batches after pmap
+        diff_JK, mask = jax.lax.fori_loop(0, batches, sequentialized_iter, [diff_JK, mask]) 
+        return [diff_JK, mask]
+
+    diff_JK, mask = jax.lax.fori_loop(0, 16, iteration, [diff_JK, mask]) 
+    return jax.lax.psum(diff_JK, axis_name="p")
+
+if __name__ == "__main__":
+    import time 
+    import argparse 
+    parser = argparse.ArgumentParser(prog='', description='', epilog='')
+    parser.add_argument('-backend', default="cpu"),
+    parser.add_argument('-natm', default=3),
+    parser.add_argument('-test', action="store_true")
+    parser.add_argument('-prof', action="store_true")
+    parser.add_argument('-batches', default=5)
+    parser.add_argument('-nipu', default=16)
+    parser.add_argument('-skip', action="store_true") 
+    args = parser.parse_args()
+    backend = args.backend 
+
+    natm = int(args.natm) 
+    nipu = int(args.nipu)
+    if backend == "cpu": nipu = 1
+
+    start = time.time()
+
+    mask = jnp.ones(1, dtype=jnp.int32)
+    mol = pyscf.gto.Mole(atom="".join(f"C 0 {1.54*j} {1.54*i};" for i in range(natm) for j in range(natm))) 
+    mol.build()
+    N = mol.nao_nr()
+    print("[%i]"%mol.nao_nr())
+    if not args.skip: 
+        print("dense_ERI", time.time()-start)
+        dense_ERI = mol.intor("int2e_sph", aosym="s1")
+    print("distinct_ERI", time.time()-start)
+    distinct_ERI = mol.intor("int2e_sph", aosym="s8")
+    print("Minao init", time.time()-start)
+    dm = pyscf.scf.hf.init_guess_by_minao(mol)         
+    HYB_B3LYP = 0.2
+    scale = HYB_B3LYP/2
+    if not args.skip: 
+        print("einsum J", time.time()-start)
+        J = np.einsum("ijkl,ji->kl", dense_ERI, dm)
+        print("einsum K", time.time()-start)
+        K = np.einsum("ijkl,jk->il", dense_ERI, dm)
+        print("J-K/2", time.time()-start)
+        truth = J - K / 2 * HYB_B3LYP
+
+    print("\n----------")
+    print("nonzero_indices ", time.time()-start)
+    nonzero_indices      = np.nonzero(distinct_ERI)[0].astype(np.int32)
+    print("grap ERI values", time.time()-start)
+    nonzero_distinct_ERI = distinct_ERI[nonzero_indices].astype(np.float32)
+    if not args.skip: 
+        print("dense: ", dense_ERI.shape, dense_ERI.nbytes/10**6)
+    print("sparse: ", distinct_ERI.shape, distinct_ERI.nbytes/10**6)
+    print("sparse+symmetric: ", nonzero_distinct_ERI.shape, nonzero_distinct_ERI.nbytes/10**6)
+    print("per ipu", nonzero_distinct_ERI.nbytes/10**6/16*2) # obs: most time might be spent compiling loading ERI from H->device! 
+    print("----------\n")
+
+    # write code that profiles get_indices/take/segment_sum independently. 
+    if args.prof: 
```

Comment:
Perhaps remove this part as it looks deprecated since some of the functions and variables no longer exist (and/or create an issue to address this?)

Reply:
Adding instead `TODO: fix broken profile code, and add profile individually for np.take and segment_sum, potentially with custom poplar implementations`